### ISSUE TITLE
How to set up the L-BFGS-B and SLSQP methods

### ISSUE THREAD

**Issue by Zyms425** [cagrikymk/JAX-ReaxFF#9]
Hello author, I am currently studying your Jax Reaxff and I am not familiar with the L-BFGS-B and SLSQP methods. I have seen that your example submission command is

jaxreaxff --init_ FF Datasets/cobalt/ffield_ lit \

--params Datasets/cobalt/params \

--geo Datasets/cobalt/geo \

--train_ file Datasets/cobalt/trainset.in \

--num_ e_ minim_ steps 200 \

--e_ minim_ LR 1e-3 \

--out_ folder ffields \

--save_ opt all \

--num_ trials 1 \

--num_ steps 20 \

--init_ FF_ type fixed \

--backend cpu

How can I use L-BFGS-B and SLSQP methods correctly? If I could get your help, I would be very grateful

**Comment by cagrikymk** [comment]
Hi,
I hardcoded the parameters for the optimizations methods (SLSQP and L-BFGS-B), so you only need to specify the optimization method using --opt_method (--opt_method L-BFGS-B or --opt_method SLSQP).

Also, I highly suggest to use the gpu backend after installing the gpu version of JAX (using --backend GPU).

**Comment by Zyms425** [comment]
Hello, excuse me.May I ask this question?


Run-Time and Training Performance Evaluation.
In JAX-ReaxFF, as mentioned above, two different gradient-
based optimization algorithms are available, L-BFGS-B and
SLSQP. For both L-BFGS-B and SLSQP, the maximum
number of iterations is set to 100 which is for the step 3 of
Algorithm 1 in Scheme 1. For L-BFGS-B, the maximum
number of iterations for the line search is set to 20, and the
maximum number of variable matrix corrections to approx-
imate the Hessian matrix is set to 20. For the rest of the control
parameters, the default values from the SciPy library are used.
The iteration count for the main optimization loop of
Algorithm 1 in Scheme 1 is set to 20 where the local error
minimization and the geometry optimization steps are
iteratively repeated this many times. Therefore, for all JAX-
ReaxFF experiments, the true error calculation with geometry
optimization is done 20 times as the local error minimization
only uses single step calculations. For experiments reported
here, the percentage of the noise used to help escaping the
local minima (line 11, Algorithm 1 in Scheme 1) is set to be
0.01%. Increasing the noise can potentially improve the results,
but this typically results in higher training times. JAX-ReaxFF
was compiled to run in single precision for all experiments
except for the disulfide case due to the numerical issues
described later. Each training experiment is repeated 10 times,
and the lowest and median training errors are reported.



Is the submission command like this：
jaxreaxff --init_FF Datasets/Co/ffield_lit&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --params Datasets/Co/params&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --geo Datasets/Co/geo&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --train_file Datasets/Co/trainset.in&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;\
	 --opt_method&nbsp; L-BFGS-B\
&nbsp; &nbsp; &nbsp; &nbsp;	 --num_trials 1\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --num_steps&nbsp; 20\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --rest_search_start&nbsp; 20\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --init_FF_type&nbsp; fixed\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--num_e_minim_steps&nbsp; &nbsp;200\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--e_minim_LR&nbsp; 1e-3\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--out_folder&nbsp; &nbsp;ffieldCo\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--save_opt&nbsp; all\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--backend&nbsp; cpu\



jaxreaxff --init_FF Datasets/Co/ffield_lit&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --params Datasets/Co/params&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --geo Datasets/Co/geo&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --train_file Datasets/Co/trainset.in&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;\
	&nbsp;--opt_method&nbsp;&nbsp;SLSQP\
&nbsp; &nbsp; &nbsp; &nbsp;	&nbsp;--num_trials 1\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --num_steps&nbsp; 20\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --rest_search_start&nbsp; 20\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; --init_FF_type&nbsp; fixed\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--num_e_minim_steps&nbsp; &nbsp;200\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--e_minim_LR&nbsp; 1e-3\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--out_folder&nbsp; &nbsp;ffieldCo\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--save_opt&nbsp; all\
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;--backend&nbsp; cpu\

I don't know if the submission command is correct。


And I am optimizing other force field processes, and the result is nan. What submission commands need to be improved to obtain force field optimization? If I could receive your answer, I would greatly appreciate it.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cagrikymk/JAX-ReaxFF"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年5月5日(星期五) 上午8:50
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [cagrikymk/JAX-ReaxFF] How to set up the L-BFGS-B and SLSQP methods (Issue #9)





 
Hi,
 I hardcoded the parameters for the optimizations methods (SLSQP and L-BFGS-B), so you only need to specify the optimization method using --opt_method (--opt_method L-BFGS-B or --opt_method SLSQP).
 
Also, I highly suggest to use the gpu backend after installing the gpu version of JAX (using --backend GPU).

**Comment by cagrikymk** [comment]
So, you were able to run it for the provided dataset but your own dataset leads to nan values?
If that is the case, first thing I suggest is to make sure your parameter ranges are okay. Also, if your starting force field is really unstable, you could run it without energy minimization to get a stable enough force field (not sure if you have data structures that require energy minimization).

If these suggestions are not feasible/does not work, you can use double precision. The current implementation uses single precision for faster convergence but I think it makes sense to change the default to double since the code already runs fast enough. I will change the default precision to double later today.

**Comment by Zyms425** [comment]
Dear author:
Hello!I would like to ask you how to obtain the gradient graph in the abstract of this paper on jax. Excuse me again, I can't find the data and method for drawing in the file，If you could provide an answer, it would be greatly appreciated
